src/contracts/Qearn.h

```diff
@@ -0,0 +1,876 @@
+                state._CurrentRoundInfo.set(qpi.epoch(), locals.updatedRoundInfo);
+                
+                output.returnCode = QEARN_LOCK_SUCCESS;          //  additional locking of this epoch is succeed
+                return ;
+            }
+
+        }
+
+        if(locals.endIndex == QEARN_MAX_LOCKS - 1) 
+        {
+            output.returnCode = QEARN_OVERFLOW_USER;
+            if(qpi.invocationReward() > 0) 
+            {
+                qpi.transfer(qpi.invocator(), qpi.invocationReward());
+            }
+            return ;                        // overflow users in Qearn
+        }
+        
+        if(qpi.invocationReward() > QEARN_MAX_LOCK_AMOUNT)
+        {
+            output.returnCode = QEARN_LIMIT_LOCK;
+            if(qpi.invocationReward() > 0) 
+            {
+                qpi.transfer(qpi.invocator(), qpi.invocationReward());
+            }
+            return;
+        }
+
+        locals.newLocker.ID = qpi.invocator();
+        locals.newLocker._Locked_Amount = qpi.invocationReward();
+        locals.newLocker._Locked_Epoch = qpi.epoch();
+
+        state.Locker.set(locals.endIndex, locals.newLocker);
+
+        locals.tmpIndex.startIndex = state.EpochIndex.get(qpi.epoch()).startIndex;
+        locals.tmpIndex.endIndex = locals.endIndex + 1;
+        state.EpochIndex.set(qpi.epoch(), locals.tmpIndex);
+
+        locals.updatedRoundInfo._Total_Locked_Amount = state._InitialRoundInfo.get(qpi.epoch())._Total_Locked_Amount + qpi.invocationReward();
+        locals.updatedRoundInfo._Epoch_Bonus_Amount = state._InitialRoundInfo.get(qpi.epoch())._Epoch_Bonus_Amount;
+        state._InitialRoundInfo.set(qpi.epoch(), locals.updatedRoundInfo);
+
+        locals.updatedRoundInfo._Total_Locked_Amount = state._CurrentRoundInfo.get(qpi.epoch())._Total_Locked_Amount + qpi.invocationReward();
+        locals.updatedRoundInfo._Epoch_Bonus_Amount = state._CurrentRoundInfo.get(qpi.epoch())._Epoch_Bonus_Amount;
+        state._CurrentRoundInfo.set(qpi.epoch(), locals.updatedRoundInfo);
+
+        output.returnCode = QEARN_LOCK_SUCCESS;            //  new locking of this epoch is succeed
+    _
+
+    struct unlock_locals {
+
+        RoundInfo updatedRoundInfo;
+        LockInfo updatedUserInfo;
+        HistoryInfo unlockerInfo;
+        
+        uint64 AmountOfUnlocking;
+        uint64 AmountOfReward;
+        uint64 AmountOfburn;
+        uint64 RewardPercent;
+        sint64 divCalcu;
+        uint32 indexOfinvocator;
+        uint32 t;
+        uint32 count_Of_last_vacancy;
+        uint32 locked_weeks;
+        uint32 startIndex;
+        uint32 endIndex;
+        
+    };
+
+    PUBLIC_PROCEDURE_WITH_LOCALS(unlock)
+
+        if(input.Locked_Epoch > QEARN_MAX_EPOCHS || input.Amount < QEARN_MINIMUM_LOCKING_AMOUNT) 
+        {
+
+            output.returnCode = QEARN_INVALID_INPUT_LOCKED_EPOCH;               //   if user try to unlock with wrong locked epoch, it should be failed to unlock.
```

Comment:
Returning `QEARN_INVALID_INPUT_LOCKED_EPOCH` if `input.Amount < QEARN_MINIMUM_LOCKING_AMOUNT` is misleading.

Reply:
the user should unlock the over 10M at once. if else, can't unlock

Reply:
Yes, I got it. so I have made new if. in this case, the returnCode should be `QEARN_INVALID_INPUT_AMOUNT`.